### PR TITLE
openai_utils.py - functionality for instantiating config_list with a .env file

### PR DESCRIPTION
**Pull request by AaronWard**
## Why are these changes needed?

I've made some enhancements to the `config_list_from_dotenv` function in the `openai_utils.py` file under the `oai` directory. These changes will allow users to load configuration details more efficiently from a `.env` file or from the environment. Here is an example:

```python
config_list = autogen.config_list_from_dotenv(
    dotenv_file_path='.env', # If None the function will try find in the working directory
    filter_dict={
        "model": {
            "gpt-4",
            "gpt-3.5-turbo",
        }
    }
)
```

Additionally, I've developed unit tests to make sure that the function behaves as expected in various scenarios, handles edge cases gracefully.

## Additional changes:
- i also added an example notebook going through each of the ways to configure your model config lists

## Related issue number

https://github.com/microsoft/autogen/issues/3

## Checks

- [ ] I've updated the documentation needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests corresponding to the changes introduced in this PR, ensuring our codebase remains robust and maintainable.
- [x] I've made sure all auto checks have passed.

Would love to hear your thoughts on this, and I'm open to any feedback or suggestions to improve it further! 😊

### CODE REVIEW

**Review by gagb — COMMENTED** [review]
Great PR!

Does `oai/__init__.py` need modifications for this to work?

**Comment by AaronWard** [comment]
> Great PR!
> 
> Does `oai/__init__.py` need modifications for this to work?

Thanks @gagb! and Good catch, i'm still making changes and cleaning up the tests, will make are ready for review when finished.

**Comment by AaronWard** [comment]
@thinkall @gagb Here is an example of the proposed code in use: As you can see there is a .env file with my openai key 

Screenshot 2023-10-02 at 10 36 59


Screenshot 2023-10-02 at 10 37 28

File: autogen/oai/openai_utils.py [hunk 1]
```diff
@@ -239,3 +243,78 @@ def config_list_from_json(
...
+
+    Raises:
+        ValueError: If no configurations are loaded or if the API key is not found in the environment variables.
+
+    Example:
+        >>> config_list_from_dotenv(dotenv_file_path='path_to_dotenv_file', api_key_env_var='OPENAI_API_KEY')
+        [
+            {'model': 'gpt-4', 'api_key': 'some_api_key'},
+            {'model': 'gpt-3.5-turbo', 'api_key': 'some_api_key'}
+        ]
```
Comment (thinkall) [inline autogen/oai/openai_utils.py:277]: Thank you @AaronWard. It looks great to me. What do you think of adding an example of the contents in the `.env` file? I'm asking because I read saw some feedbacks mentioning not knowing what contents should be in the configure files used in the example notebooks.

Comment (AaronWard) [reply autogen/oai/openai_utils.py:277]: @thinkall Sure, i will make an instructions notebook in [`notebooks directory`](https://github.com/microsoft/autogen/tree/main/notebook) with the different ways of setting up the configurations using:
- `config_list_gpt4_gpt35`
- `config_list_openai_aoai`
- `config_list_from_models`
- `config_list_from_json`

I'll try be as explicit as possible because i think the JSON structure requirement seems to catch people off-guard, particularly for novice developers. I'll get this done later today.

Comment (thinkall) [reply autogen/oai/openai_utils.py:277]: Thank you @AaronWard, would you like to add the notebook in this PR?

Also, some tests failed, could you check about it?

Comment (AaronWard) [reply autogen/oai/openai_utils.py:277]: Working on it now 👍

**Comment by thinkall** [comment]
> @sonichi @thinkall @gagb
> 
> I've added a notebook going over all the says to set up a `config_list`. please review and let me know if anything is incorrect.
> 
> https://github.com/microsoft/autogen/blob/d52967261752aa7ce4954d134b8df1a298f1cb7d/notebook/oai_openai_utils.ipynb

Thank you @AaronWard , the notebook looks good to me. But the tests are failed, I see some permission errors and some assertion errors.

**Comment by sonichi** [comment]
The tests fail on windows: https://github.com/microsoft/autogen/actions/runs/6406254119/job/17390491641?pr=68
I suppose it's because the line of loading the temp file is inside the `with` statement.

**Review by thinkall — APPROVED** [review]
Thanks a lot @AaronWard , approved.
